rfd/0180-sso-as-mfa-method.md

```diff
@@ -0,0 +1,528 @@
+Currently, privilege tokens can only be used in select endpoints as a replacement
+for MFA. Privilege Tokens will be supported as a normal MFA response to support
+other use cases:
+
+```proto
+message MFAAuthenticateResponse {
+  oneof Response {
+    TOTPResponse TOTP = 2;
+    webauthn.CredentialAssertionResponse Webauthn = 3;
+    PrivilegeTokenResponse Token = 4;
+  }
+}
+```
+
+Note: Privilege tokens used in this way must be consumed upon verification to
+ensure that exchanging a webauthn or totp token for a privilege token through
+`rpc CreatePrivilegeToken` has no significant security implications.
+
+#### SSO MFA flow
+
+We will introduce a new SSO login flow that allows clients to get a new privilege
+token instead of the standard login credentials. Authenticated clients can create
+an sso auth request with `rpc CreateSAMLAuthRequest`, `rpc CreateOIDCAuthRequest`
+or `rpc CreateGithubAuthRequest` and set `CreatePrivilegeToken=true` to use this
+new flow. Once the client authenticates through the IdP, the client will be given
+a privilege token generated for the user.
+
+Note: As is the case with normal SSO login, the login response is encrypted
+using a secret key owned by the client so that the token can not be intercepted
+in a man in the middle attack.
+
+#### Forced Re-authentication
+
+By default, Teleport will require the user to re-login through their IdP to
+pass an MFA check. This will prevent long lasting login sessions from being
+used as MFA verification, which could easily be exploited by an attacker with
+remote access to the user's machine.
+
+For SAML, this can be achieved by setting `ForceAuthn=true` in the authn
+request. For OIDC and Github, setting `max_age=0` in the server redirect URL's
+query parameter will force re-auth.
+
+In cases where re-auth is not desired behaviour, such as when the IdP is
+configured to prompt for MFA for existing sessions, users can set `force_authn`
+or `max_age` in the SAML or OIDC connector respectively to override the default
+behavior.
+
+### Proto
+
+**MFAConnectorName**
+```diff
+message AuthPreferenceSpecV2 {
+  // Type is the type of authentication.
+  string Type = 1 [(gogoproto.jsontag) = "type"];
+  ...
+  // ConnectorName is the name of the OIDC or SAML connector. If this value is
+  // not set the first connector in the backend will be used.
+  string ConnectorName = 3 [(gogoproto.jsontag) = "connector_name,omitempty"];
+  ...
++  // MFAConnectorName is the name of an auth connector to use for MFA verification.
++  // If this value is not set, the first connector in the backend with AllowAsMFAMethod
++  // set to YES or ONLY will be used, starting with ConnectorName.
++  string MFAConnectorName = 21 [(gogoproto.jsontag) = "mfa_connector_name,omitempty"];
++  // MFAConnectorType is the type of auth connector to use for MFA verification, if any.
++  // Defaults to the auth Type set above.
++  string MFAConnectorType = 22 [(gogoproto.jsontag) = "mfa_connector_type,omitempty"];
```

Comment:
Isn't this redundant given that we have MFAConnectorName?

What if MFAConnectorName and MFAConnectorType disagree?

My 2c is that I would cut this toggle until someone asks for it (and makes a good case while asking).